pmm-app/src/pmm-qan-app-panel/QueryAnalytics/StateContext.tsx

```diff
@@ -0,0 +1,279 @@
+import React, { useEffect, useState } from 'react';
+import { ParseQueryParamDate } from '../../react-plugins-deps/components/helpers/time-parameters-parser';
+import { getDataSourceSrv } from '@grafana/runtime';
+import _ from 'lodash';
+
+const initialState = {} as any;
+
+export const StateContext = React.createContext(initialState);
+
+const DEFAULT_COLUMNS = ['load', 'num_queries', 'query_time'];
+const FILTERS_NAMES = [
+  'environment',
+  'cluster',
+  'replication_set',
+  'database',
+  'schema',
+  'node_name',
+  'service_name',
+  'client_host',
+  'username',
+  'service_type',
+  'node_type',
+  'city',
+  'az',
+];
+
+class ContextActions {
+  constructor() {}
+
+  static setFilters(query) {
+    return FILTERS_NAMES.reduce((result, filterName) => {
+      const filters = query.getAll(`var-${filterName}`);
+      if (!filters.length) {
+        return result;
+      }
+
+      if (filters[0] === 'All' || filters[0] === '') {
+        return result;
+      }
+
+      result[filterName] = filters;
+      return result;
+    }, {});
+  }
+
+  static refreshGrafanaVariables(state) {
+    const dataSource = getDataSourceSrv();
+    // @ts-ignore
+    const templateVariables = dataSource.templateSrv.variables;
+    FILTERS_NAMES.forEach(filter => {
+      const variables = _.find(templateVariables, { name: filter.replace('var-', '') });
+      if (!variables) {
+        return;
+      }
+      variables.current = {
+        text: state.labels[filter] || ['All'],
+        value: state.labels[filter] || ['All'],
+      };
+    });
+    templateVariables[0].variableSrv.variableUpdated(templateVariables[0]);
+  }
+
+  static generateURL(state) {
+    // read parameters and create new url
+    // @ts-ignore
+    const labels =
+      state.labels &&
+      Object.keys(state.labels)
+        .map(key => {
+          // @ts-ignore
+          const variables = state.labels[key];
+          return variables.map(variable => `var-${key}=${variable}`).join('&');
+        })
+        .filter(Boolean)
+        .join('&');
+    const columnsQuery = state.columns ? `columns=${JSON.stringify(state.columns)}` : '';
+    const groupBy = state.groupBy ? `group_by=${state.groupBy}` : '';
+    const filterByQuery = state.queryId ? `filter_by=${state.queryId}` : '';
+    const orderBy = state.orderBy ? `order_by=${state.orderBy}` : '';
+    const from = state.rawTime && state.rawTime.from ? `from=${state.rawTime.from}` : '';
+    const to = state.rawTime && state.rawTime.to ? `to=${state.rawTime.to}` : '';
+    // TODO: replace crutch with right redirect
+    return `${window.location.pathname}?${[columnsQuery, filterByQuery, labels, orderBy, groupBy, from, to]
```

Comment:
Feels like too many things in the `return` statement, difficult to read and comprehend what is returned.